src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -0,0 +1,172 @@
+        // multiple sub queries, doc ids may repeat for each sub query results
+        return queryTopDocs.stream()
+            .map(compoundQueryTopDocs -> combineShardScores(scoreCombinationTechnique, compoundQueryTopDocs))
+            .collect(Collectors.toList());
+    }
+
+    @VisibleForTesting
+    protected float combineShardScores(
+        final ScoreCombinationTechnique scoreCombinationTechnique,
+        final CompoundTopDocs compoundQueryTopDocs
+    ) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+            return ZERO_SCORE;
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
+
+        // - create map of combined scores per doc id
+        Map<Integer, Float> combinedNormalizedScoresByDocId = combineScoresAndGetCombinedNormalizedScoresPerDocument(
+            normalizedScoresPerDoc,
+            scoreCombinationTechnique
+        );
+
+        // - sort documents by scores and take first "max number" of docs
+        // create a collection of doc ids that are sorted by their combined scores
+        List<Integer> sortedDocsIds = getSortedDocIds(combinedNormalizedScoresByDocId);
+
+        // - update query search results with normalized scores
+        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, sortedDocsIds);
+
+        // return max score
+        if (sortedDocsIds.isEmpty()) {
+            return ZERO_SCORE;
+        }
+        return combinedNormalizedScoresByDocId.get(sortedDocsIds.get(0));
+    }
+
+    @VisibleForTesting
+    protected List<Integer> getSortedDocIds(final Map<Integer, Float> combinedNormalizedScoresByDocId) {
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
+        sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
+        return sortedDocsIds;
+    }
+
+    @VisibleForTesting
+    protected ScoreDoc[] getCombinedScoreDocs(
+        final CompoundTopDocs compoundQueryTopDocs,
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final List<Integer> sortedScores,
+        final int maxHits
+    ) {
+        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits && j < sortedScores.size(); j++) {
+            int docId = sortedScores.get(j);
+            finalScoreDocs[j] = new ScoreDoc(docId, combinedNormalizedScoresByDocId.get(docId), shardId);
```

Comment:
we are not updating the shard id? why?

Reply:
at this step we're working with a single shard results. It's the same shard id, so we need just to copy it to new results

Reply:
It would be useful if you can add the comments in the code, around which operation is at shard level and what is not.